### ISSUE TITLE
add sysban type: OpenID login source

### ISSUE THREAD

**Issue by rahaeli** [dreamwidth/dreamwidth#3353]
While trying to solve a problem where this could be useful I discovered we don't actually have this sysban type and we should: there should be a way to sysban OpenID providers by URL and if a provider is on that sysban list, nobody should be able to log into or comment using an OpenID from that host.

**Comment by pauamma** [comment]
Does this need wildcards? That is, do you foresee a need to give partial URLs, like "https://*.example.org/" or "https://www.example.net/users/*" for everyone from example.org/example.net, or is it enough for your purpose to have only literal matching like "https://exampleuser1.example.org/" and "https://www.example.net/users/exampleuser2", for these specific users?

**Comment by cmho** [comment]
yeah, i think based on the described case we'd want wildcard matching since the idea is to be able to ban a whole host from openid

**Comment by rahaeli** [comment]
Yeah, top-level domain matching is probably the best (ie, enter example.org and it matches any subdomain)

**Comment by kareila** [comment]
We already have email_domain for account creation, let’s call it openid_domain?

**Comment by pauamma** [comment]
Speaking of email_domain sysbans, https://github.com/dreamwidth/dreamwidth/blob/main/cgi-bin/LJ/Sysban.pm#L261 shows there's 1-level wildcarding, so an email_domain sysban on example.com also bans third-level.example.com, but not fourth-level.third-level.example.com or deeper subdomains. Is that deep enough for your purpose, or do you also want an example.com openid_domain sysban to affect all subdomains no matter their depth?

**Comment by rahaeli** [comment]
Matching the behavior of email_domain is fine!

**Comment by pauamma** [comment]
Last set of questions, I hope:

- What message should be displayed to the user at OpenID login time if the entered URL matches an openid_domain sysban?
- What (if anything) should happen to existing OpenID sessions that match an openid_domain sysban added after login? Easiest (if they shouldn't still be allowed) is probably forcibly logging them out with no message when the next request comes in, somewhere in whatever request processing stage extracts the logged-in user from cookies. (Pointer to the best place to do it welcome, so the check also applies to the WIP API if at all possible.)

**Comment by rahaeli** [comment]
- "We're sorry, but [[sitename]] has restricted OpenID logins from your OpenID provider. Please contact us if you believe this restriction is in error."
- Forcible logout is fine. (I can't answer the "best place" question, alas!)

**Comment by pauamma** [comment]
>     * Forcible logout is fine. (I can't answer the "best place" question, alas!)

@momijizukamori maybe?

**Comment by momijizukamori** [comment]
So the main login logic is in https://github.com/dreamwidth/dreamwidth/blob/main/cgi-bin/LJ/User/Login.pm buuuut I'm not sure that gets hit for things other than transitioning from logged out to logged in. But there may be something in there you can hook into for auth checks (it looks like `LJ::get_remote()` is what the new-style pages use in the controller)

The new API uses a totally different auth mechanism (API keys) than session cookies, and I don't think OpenID users get access to them though uh I should probably double-check that. (though all the API routes use a user object fetched in reference to the key, and do all the usual 'can this user update/view that resource' checks we do with cookie sessions, so there's no security risks)

**Comment by pauamma** [comment]
Thanks.
